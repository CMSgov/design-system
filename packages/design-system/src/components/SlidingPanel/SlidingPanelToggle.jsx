import Button from '../Button/Button';
import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

/**
 * A link that triggers the visibility of a sliding panel
 */
class SlidingPanelToggle extends React.PureComponent {
  componentDidUpdate(prevProps) {
    if (!this.props.panelOpen && prevProps.panelOpen && this.buttonRef) {
      this.buttonRef.focus();
    }
  }

  render() {
    const { className, children, inline, showPanel, panelOpen, ...others } = this.props;
    const classes = classNames(
      'ds-c-sliding-panel__toggle',
      inline && 'ds-c-sliding-panel__toggle--inline',
      className
    );

    return (
      <Button
        className={classes}
        inputRef={(el) => (this.buttonRef = el)}
        onClick={showPanel}
        variation="transparent"
        {...others}
      >
        {children}
      </Button>
    );
  }
}

SlidingPanelToggle.propTypes = {
  /**
   * Determines if SlidingPanel is open or closed.
   * This value is used to re-focus the toggle that opened the panel when the panel closes.
   */
  panelOpen: PropTypes.bool.isRequired,
  /**
   * SlidingPanelToggle content.
   */
  children: PropTypes.node.isRequired,
  /**
   * Additional classes for the toggle button anchor element.
   */
  className: PropTypes.string,
  /**
   * Adds `display: inline` to the SlidingPanelToggle.
   */
  inline: PropTypes.bool,
  /**
   * This function is called with an id that the toggle generates.
   * It can be used in implementing the SlidingPanel for keeping track of which panel the toggle controls.
   */
  showPanel: PropTypes.func.isRequired,
};

export default SlidingPanelToggle;
