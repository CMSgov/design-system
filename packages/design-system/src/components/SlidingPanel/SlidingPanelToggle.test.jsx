import SlidingPanelToggle from './SlidingPanelToggle.jsx';
import React from 'react';
import { shallow } from 'enzyme';

const defaultProps = {
  panelOpen: false,
  inline: false,
  showPanel: jest.fn(),
};

function renderSlidingPanelToggle(props) {
  props = Object.assign({}, defaultProps, props);
  const wrapper = shallow(
    <SlidingPanelToggle {...props}>
      <p>content</p>
    </SlidingPanelToggle>
  );
  return { props, wrapper };
}

describe('SlidingPanelToggle', () => {
  beforeEach(() => {
    defaultProps.showPanel.mockClear();
  });

  it('renders a button', () => {
    const { wrapper } = renderSlidingPanelToggle();
    expect(wrapper).toMatchSnapshot();
  });

  it('calls showPanel() on toggle click', () => {
    const { wrapper } = renderSlidingPanelToggle();
    const toggle = wrapper.find('Button');
    toggle.simulate('click');
    expect(defaultProps.showPanel).toHaveBeenCalled();
  });

  it('applies inline display style via inline prop', () => {
    const { wrapper } = renderSlidingPanelToggle({ inline: true });
    const toggle = wrapper.find('Button');
    expect(toggle.hasClass('ds-c-sliding-panel__toggle--inline')).toBe(true);
  });

  it('applies custom class via className prop', () => {
    const className = 'test-class';
    const { wrapper } = renderSlidingPanelToggle({ className });
    const toggle = wrapper.find('Button');
    expect(toggle.props().className).toContain(className);
  });

  it('passes through extra props', () => {
    const ariaLabel = 'test';
    const { wrapper } = renderSlidingPanelToggle({ ariaLabel });
    const toggle = wrapper.find('Button');
    expect(toggle.props().ariaLabel).toBe(ariaLabel);
  });
});
