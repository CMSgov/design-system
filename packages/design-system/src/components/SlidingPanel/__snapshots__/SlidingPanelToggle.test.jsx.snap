// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlidingPanelToggle renders a button 1`] = `
<Button
  className="ds-c-sliding-panel__toggle"
  component="button"
  inputRef={[Function]}
  onClick={[MockFunction]}
  type="button"
  variation="transparent"
>
  <p>
    content
  </p>
</Button>
`;
